Concurrency: swift_task_startOnMainActor

etcwilde opened the pull request:
This patch adds an SPI to run the first partial function of a MainActor asynchronous function on the MainActor synchronously. This is effectively like the asynchronous program entrypoint behavior. The first partial function is run synchronously. Following continuations are enqueued for execution like any other asynchronous function.

etcwilde committed:
Concurrency: Task.startOnMainActor

This patch adds an SPI to run the first partial function of a MainActor
asynchronous function on the MainActor synchronously. This is
effectively like the asynchronous program entrypoint behavior. The first
partial function is run synchronously. Following continuations are
enqueued for execution like any other asynchronous function.

DougGregor commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -186,6 +186,24 @@ extension Task: Equatable {
 }
 
+@available(SwiftStdlib 5.9, *)
+extension Task where Failure == Never {
```
Let's add the `Failure == Error` case as well for throwing work

DougGregor commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -186,6 +186,24 @@ extension Task: Equatable {
+        _ work: @escaping @MainActor() async -> Success
+    ) -> Task<Success, Never> {
+        let flags = taskCreateFlags(priority: priority, isChildTask: false,
+                                    copyTaskLocals: false, inheritContext: false,
```
Hmm. I think this probably should inherit context, because it's meant for main-thread work?

ktoso replied:
Yeah, I think this needs the `@_unsafeInheritExecutor` treatment

DougGregor commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -1624,5 +1624,13 @@ static void swift_task_asyncMainDrainQueueImpl() {
+  AsyncTask * originalTask = swift_task_getCurrent();
+  ExecutorRef mainExecutor = swift_task_getMainExecutor();
+  swift_job_run(job, mainExecutor);
+  _swift_task_setCurrent(originalTask);
```
Shouldn't this runtime function take in a task and convert it to a job?

etcwilde replied:
Sure, fixed the function decl. `AsyncTask` is a `Job`, and `convertTaskToJob` looks like it's just a bitcast so I don't think I need to do anything more there.

etcwilde commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -186,6 +186,24 @@ extension Task: Equatable {
 }
 
+@available(SwiftStdlib 5.9, *)
+extension Task where Failure == Never {
```
With the appropriate `throws` and `rethrows` annotations, this might just work with throwing functions too? I haven't tested that yet though.

DougGregor replied:
Yes

etcwilde commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -186,6 +186,24 @@ extension Task: Equatable {
+extension Task where Failure == Never {
+    @_spi(MainActorUtilities)
+    @MainActor
+    static func taskStartingOnMainActor(
```
heh, I thought I'd done that. Yeah, good call.

mikeash commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -1594,6 +1594,15 @@ void (*swift::swift_task_asyncMainDrainQueue_hook)(
 
+SWIFT_CC(swift)
+void swift::swift_task_startOnMainActor(AsyncTask* task) {
+  AsyncTask * originalTask = swift_task_getCurrent();
```
Can we assert that we are actually on the main actor here?

ktoso replied:
You might like these: https://github.com/apple/swift/pull/64062 if you wanted to assert still in swift. but yeah, maybe good to do so here in C++

rokhinip replied:
asserts only run on debug builds right? Do we want to actually check this and then raise a fatalError?

rokhinip replied:
I think you might want to swift_task_clearCurrent() instead though what you have works since you end up setting a task on the thread in swift_job_run() anyhow.

ktoso replied:
Maybe always check and fatalError like @rokhinip suggested? You say @etcwilde that it's only surfaced in a @MainActor API but still, if we were to use it wrongly it'd be good to realize... 🤔  if it shows up in profiling a lot we could downgrade to an assertion perhaps.

etcwilde replied:
> asserts only run on debug builds right? Do we want to actually check this and then raise a fatalError?

Yes. This would only show up in an asserts build of the runtime. Since it's the runtime under SPI, it's pretty buried. The only way someone would expose this is if we added a new API/usage that doesn't adhere to this requirement. In theory, CI is running Asserts builds and should catch it.

> I think you might want to swift_task_clearCurrent() instead though what you have works since you end up setting a task on the thread in swift_job_run() anyhow.

Where should I be calling that? The `swift_task_getCurrent()` returns `nullptr` after coming back which is causing crashes in `swift_task_wait_futureImpl`

```c++
...
  TaskContinuationFunction *resumeFn,
  AsyncContext *callContext) {
  // Suspend the waiting task.
  auto waitingTask = swift_task_getCurrent();
  waitingTask->ResumeTask = task_future_wait_resume_adapter; // boom! waiting task is nullptr
  waitingTask->ResumeContext = callContext;
...
```

mikeash replied:
Getting the current executor should be cheap, so I vote for checking unconditionally to start with, and downgrading if it turns out to be a performance problem somehow.

etcwilde replied:
Fatal error added.

mikeash commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -1594,6 +1594,15 @@ void (*swift::swift_task_asyncMainDrainQueue_hook)(
     swift_task_asyncMainDrainQueue_override compatOverride) = nullptr;
 
+SWIFT_CC(swift)
+void swift::swift_task_startOnMainActor(AsyncTask* task) {
```
Should this function be overrideable? Seems very basic, not much need for it, but....

etcwilde replied:
I... maybe? I wasn't sure. What's your rule-of-thumb for overriding things? I think the flags are the part I'd be most worried about, but those are generated in Swift.

etcwilde replied:
Okay, it's overridable now.

harlanhaskins commented:
One thing that is still kind of an issue here when dealing with legacy concurrency is kicking off concurrent work synchronously and expecting it to finish later, which I don’t think this addresses.

The example I can think of is NSItemProvider and UIDropInteraction, which has internal bookkeeping to keep the session alive if you’ve kicked off an item load in performDrop. Without a way to synchronously construct a continuation, we still can’t use this…

```swift
Task.startOnMainActor {
    // This is gonna introduce a suspension point, defeating the purpose of the start-synchronously behavior
    await withCheckedThrowingContinuation { c in
        item.load { url, error in
            c.resume(returning: url)
        }
    }
}
```

ktoso commented:
> creating continuations synchronously, which I don’t think this addresses.

Hm? Creating continuations is synchronous; I just corrected our documentation about those even: https://github.com/apple/swift/pull/64050

etcwilde commented:
@swift-ci please test

ktoso commented:
> ....oh, really? Wow I've been operating on the incorrect assumption that we'd need two APIs to solve this problem. Okay, I'll give this a shot once I can.

Yeah I can see how one can be led to this conclusion: 1) our docs lied 2) we train people to think "await == suspension point" but the withContinuation APIs are a little bit special in that respect. Glad you chimed in so we could clear that up @harlanhaskins !

ktoso reviewed (approved):
Not in love with it being a bit adhoc but the implementation seems correct 👍

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -186,6 +186,42 @@ extension Task: Equatable {
+        _ work: @escaping @MainActor() async throws -> Success
+    ) -> Task<Success, Error> {
+        let flags = taskCreateFlags(priority: priority, isChildTask: false,
+                                    copyTaskLocals: false, inheritContext: true,
```
oh no, wait, the `copyTaskLocals` should be `true` here; we're semantically continuing work so we should not "effectively _detach_" here

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -186,6 +186,42 @@ extension Task: Equatable {
+        _ work: @escaping @MainActor() async -> Success
+    ) -> Task<Success, Never> {
+        let flags = taskCreateFlags(priority: priority, isChildTask: false,
+                                    copyTaskLocals: false, inheritContext: true,
```
This also should be true
```suggestion
                                    copyTaskLocals: true, inheritContext: true,
```

etcwilde committed:
Concurrency: Add startOnMainActor test

Adding a test to ensure that we're running things while the main actor
has other things enqueued.

etcwilde committed:
Concurrency: appease the api-digester

Looks like simply putting the availability attribute on the extension
does not suffice. Putting the availability attr on the function itself
to appease the api-digester.

etcwilde commented:
@swift-ci please test

etcwilde commented:
Alright, for attributes on the incoming function, do we want `@_ImplicitSelfCapture`, `@_inheritAsyncContext` and maybe an `__owned @Sendable` perhaps?

etcwilde commented:
Alright, edits force-pushed:
- Made throwing extension condition `extension Task where Failure == Error {`, fixing the inference issue.
- Made incoming closure `@_inheritActorContext @_implicitSelfCapture _ work: __owned @Sendable @escaping @MainActor() async {throws} -> Success`
- Added test for throwing case

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test macOS

ktoso reviewed (approved):
Re-reviewed, additions look good. The annotations on the func also, thank you @etcwilde !

DougGregor reviewed (approved):
LGTM
